### PR TITLE
Rework cbuf to use FreeRTOS Ringbuffer

### PR DESCRIPTION
**Pull request by me-no-dev**
- [x] Implement `cbuf::peek`

### CODE REVIEW

**Comment by VojtechBartoska** [comment]
@me-no-dev is this ready for review?

**Comment by me-no-dev** [comment]
yes. pending because only ugly ideas on how to implement it at the moment. Maybe worth checking if anything that uses it, actually calls `peek`

**Comment by lucasssvaz** [comment]
@me-no-dev From what I saw it is only used by the WiFi and there's no use of the `peek()` method. I think it can be left as it is for now.

**Comment by lucasssvaz** [comment]
Crashing in `cbuf::write()`. Tested using the BasicOTA example.

```
Decoding stack results
0x4002f7fb: vListInsert at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/FreeRTOS-Kernel/list.c line 171
0x4002e519: vTaskPlaceOnEventList at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/FreeRTOS-Kernel/tasks.c line 3799
0x4002d7ce: xQueueSemaphoreTake at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/FreeRTOS-Kernel/queue.c line 1839
0x4002d854: xQueueTakeMutexRecursive at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/FreeRTOS-Kernel/queue.c line 783
0x40087bc1: cbuf::write(char const*, unsigned int) at /home/lucassvaz/Arduino/hardware/espressif/esp32/cores/esp32/cbuf.cpp line 227
0x400847ff: WiFiUDP::parsePacket() at /home/lucassvaz/Arduino/hardware/espressif/esp32/libraries/WiFi/src/WiFiUdp.cpp line 228
0x40085822: ArduinoOTAClass::handle() at /home/lucassvaz/Arduino/hardware/espressif/esp32/libraries/ArduinoOTA/src/ArduinoOTA.cpp line 379
0x40082bfa: loop() at /home/lucassvaz/Arduino/sketches/test/test.ino line 67
0x40087c88: loopTask(void*) at /home/lucassvaz/Arduino/hardware/espressif/esp32/cores/esp32/main.cpp line 74
```

**Comment by me-no-dev** [comment]
@lucasssvaz i kinda wonder why it's calling recursive api... can you try without?

**Comment by me-no-dev** [comment]
and what does it give as reason?

**Comment by me-no-dev** [comment]
@lucasssvaz can you make WiFiUdp not use cbuf, so we can delete the class altogether?

**Comment by lucasssvaz** [comment]
> @lucasssvaz can you make WiFiUdp not use cbuf, so we can delete the class altogether?

I'll take a look

**Comment by lucasssvaz** [comment]
@me-no-dev Fixed crashes and "implemented" the peek function (it is indeed required). Now BasicOTA works as expected. 

The implementation of the peek function is a very makeshift solution that requires copying the `Ringbuffer_t` structure from IDF. Idk if it is an acceptable solution. If it is, the correct way to do it would be to extract the required info into a separate header file using the lib-builder. Something like this would do the job:

```bash
echo "#pragma once" > ringbuffer_def.h
grep "typedef struct RingbufferDefinition" esp_ringbuf/ringbuf.c  >> ringbuffer_def.h
grep "typedef BaseType_t (*CheckItemFitsFunction_t)" esp_ringbuf/ringbuf.c  >> ringbuffer_def.h
grep "typedef void (*CopyItemFunction_t)" esp_ringbuf/ringbuf.c  >> ringbuffer_def.h
grep "typedef void *(*GetItemFunction_t)" esp_ringbuf/ringbuf.c  >> ringbuffer_def.h
grep "typedef void (*ReturnItemFunction_t)" esp_ringbuf/ringbuf.c  >> ringbuffer_def.h
grep "typedef size_t (*GetCurMaxSizeFunction_t)" esp_ringbuf/ringbuf.c  >> ringbuffer_def.h
awk '/typedef struct RingbufferDefinition {/,/} Ringbuffer_t;/' esp_ringbuf/ringbuf.c  >> ringbuffer_def.h
```

**Comment by lucasssvaz** [comment]
@me-no-dev Properly implemented as we discussed.

**Review by lucasssvaz — APPROVED** [review]

File: cores/esp32/cbuf.cpp [hunk 1]
```diff
@@ -34,7 +34,9 @@
 #endif
 
 cbuf::cbuf(size_t size) :
-    next(NULL), _buf(xRingbufferCreate(size, RINGBUF_TYPE_BYTEBUF))
+    next(NULL),
+    has_peek(false),
```
Comment (me-no-dev) [inline cores/esp32/cbuf.cpp:38]: also `peek_byte(0),` here

**Pull request merged**